pkg/yttlibrary/all.go

```diff
@@ -7,18 +7,34 @@ import (
 	"fmt"
 
 	"github.com/k14s/starlark-go/starlark"
+	"github.com/k14s/starlark-go/starlarkstruct"
 	tplcore "github.com/vmware-tanzu/carvel-ytt/pkg/template/core"
 	"github.com/vmware-tanzu/carvel-ytt/pkg/yttlibrary/overlay"
 )
 
+var registeredExts []*starlarkstruct.Module
+
+// DoNotUseRegisterExt allows to registry ytt library extensions.
+// It's important to allow only specific module names to be registered
+// as we do not want end-users of ytt to extend @ytt library.
+// (They should define their own library for that.)
+func DoNotUseRegisterExt(mod *starlarkstruct.Module) {
```

Comment:
```suggestion
// RegisterExt adds "mod" to the standard set of ytt library modules as an extension.
// An "extension" is a Starlark module that has external Go dependencies
// (as opposed to other Starlark modules in the ytt library that either have no dependencies or depend on the Go standard lib)
// This enables those using ytt as a Go module to opt-in (rather than be forced) to accept such dependencies.
// Only Carvel-maintained extensions can be registered; this reserves the `@ytt:` namespace.
// Integrators who want to write their own extensions should construct their own library.
func RegisterExt(mod *starlarkstruct.Module) {
```

Given that it contains mechanism that implements the policy, it feels a bit "overkill" to _also_ prepend the name of the function with `DoNotUse...`.
Add explicit definition of what an "extension" is; providing a well-defined concept for this overall mechanism (registering).
Also suggesting additional context for usage by maintainers.

Reply:
@cppforlife noted:
> the reason why ive added DoNotUse is because i want to strongly discourage anybody from using it.
> it's really a "private" method to ytt project, but go doesnt have a way to define it.

Yes. The menu of "channels" to "signal" here include:
- the method name
- Go Doc
- error messages.

I'm suggesting that the latter two are sufficient to send this signal.
Further that we're "denormalizing" the method namespace with this kind of prefix. (tho, TBH, if we would adopt something like `ModulePrivate_`, `Private` or just `_` as a prefix, I'd be _more_ okay with it — such a convention would easily apply for both integrators and contributors without (much) of a story to tell).

Reply:
We spoke about this.

For now, we'll adopt the changes I suggested.

At the same time, @cppforlife has some best-laid plans for providing for a proper library extensions feature for ytt. So, this will likely evolve through that work.